File: Algo_Ds_Notes/Exponential_Search/Exponential_Search.java
import java.util.*; 

class Exponential_Search 
{ 
    // Returns position of first occurrence of x in array 
    static int expoSearch(int input[], int n, int x) 
    { 
        // If x is present at firt location itself 
	if (input[0] == x) 
	    return 0; 
	
	// Find range for binary search by repeated doubling 
	int i = 1; 
	while (i < n && input[i] <= x) 
	    i = i*2; 
	
	// Call binary search for the found range. 
	return Arrays.binSearch(input, i/2, Math.min(i, n), x); 
    } 
	
    // Driver code 
    public static void main(String args[]) 
    { 
	
	int n;
	Scanner s = new Scanner(System.in);
	n = s.nextInt();
        int arr[] = new int[n];
	for (int i = 0; i < n; i++)
        {
            a[i] = s.nextInt();
        }
	int x = s.nextInt();
	int result = expoSearch(arr, arr.length, x); 
		
	System.out.println((result < 0) ? "Element is not present in array" : "index " + result); 
     } 
} 

/* 
INPUT:
n = 5
arr[] = {2, 3, 14, 20, 40}
x = 20
OUPUT:
index 3
*/

File: Algo_Ds_Notes/Exponential_Search/Exponential_Search.py

def binSearch( arr, left, right, x): 
    
    if right >= left: 
        mid = left + ( right-left ) / 2
          
        # If the element is present at the middle itself 
        if arr[mid] == x: 
            return mid 
          
        # If the element is smaller than mid,then it can only be present in the left subarray 
        if arr[mid] > x: 
            return binSearch(arr, left, mid - 1, x) 
          
        # Else he element can only be present in the right 
        return binSearch(arr, mid + 1, right, x) 
        
    return -1



def exponentialSearch(arr, n, x): 

    # IF x is present at first  location itself 
    if arr[0] == x: 
        return 0
          
    # Find range for binary search j by repeated doubling 
    i = 1
    while i < n and arr[i] <= x: 
        i = i * 2
      
    # Call binary search for the found range 
    return binSearch( arr, i / 2,  
                         min(i, n), x) 
      
  
# Driver Code 
n = int(input())
arr = []
for i in range(n):
    x = int(input())
    arr.append(x) 
x = int(input())       
final = exponentialSearch(arr, n, x) 
if final == -1: 
    print ("Element not found in thye array")
else: 
    print ("index %d" %(final)) 

'''
INPUT:
n = 5
arr[] = {2, 3, 14, 20, 40}
x = 20
OUPUT:
index 3
'''
